pkg/platform/update-api/update_api.go

```diff
@@ -0,0 +1,134 @@
+	Ready     updateState = "Ready"
+)
+
+type UpdateImage struct {
+	Arch    string `json:"arch"`
+	Version string `json:"version"`
+	Variant string `json:"variant"`
+}
+
+func (ui *UpdateImage) Identifier() interface{} {
+	return ui.Version
+}
+
+type StagedImage struct {
+	Image      UpdateImage `json:"image"`
+	NextToBoot bool        `json:"next_to_boot"`
+}
+
+type updateCommand = string
+
+const (
+	refresh    updateCommand = "refresh"
+	prepare    updateCommand = "prepare"
+	activate   updateCommand = "activate"
+	deactivate updateCommand = "deactivate"
+)
+
+type commandStatus = string
+
+const (
+	Success commandStatus = "Success"
+	Failed  commandStatus = "Failed"
+	Unknown commandStatus = "Unknown"
+)
+
+type CommandResult struct {
+	CmdType    updateCommand `json:"cmd_type"`
+	CmdStatus  commandStatus `json:"cmd_status"`
+	Timestamp  string        `json:"timestamp"`
+	ExitStatus *int32        `json:"exit_status"`
+	Stderr     *string       `json:"stderr"`
+}
+
+type UpdateStatus struct {
+	UpdateState       updateState    `json:"update_state"`
+	AvailableUpdates  []string       `json:"available_updates"`
+	ChosenUpdate      *UpdateImage   `json:"chosen_update"`
+	ActivePartition   *StagedImage   `json:"active_partition"`
+	StagingPartition  *StagedImage   `json:"staging_partition"`
+	MostRecentCommand *CommandResult `json:"most_recent_command"`
+}
+
+// getUpdateStatus repeatedly tries to GET /updates/status until a status is returned
+// We only retry the request if the response was a 423 Locked (indicates that the update status is not ready)
+func getUpdateStatus(httpClient http.Client) (*UpdateStatus, error) {
+	var response *http.Response
+	var err error
+	var attempts = 0
+	// Retry up to 5 times in case the Update API is busy; Waiting 10 seconds between each attempt.
+	for attempts < 5 {
+		response, err = httpClient.Get("http://unix" + "/updates/status")
+		if err != nil {
+			errors.New("error when trying to GET '/updates/status'")
+			return nil, err
+		}
+		// Response OK, break out of loop
+		if response.StatusCode == 204 {
+			break
+		} else if response.StatusCode == 423 {
+			// Retry if we get a 423 Locked response (update API busy)
+			time.Sleep(10 * time.Second)
+		} else {
+			return nil, errors.Errorf("bad http response when trying to GET '/actions/refresh-updates': status code %d", response.StatusCode)
+		}
+	}
+	if attempts == 5 {
+		return nil, errors.New("ran out of tries waiting for the Update API to be available")
+	}
+
+	var updateStatus *UpdateStatus
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(body, updateStatus)
+	if err != nil {
+		return nil, err
+	}
+	return updateStatus, nil
+}
+
+func postAction(httpClient http.Client, path string) error {
+	var response *http.Response
+	var err error
+	var attempts = 0
+	// Retry up to 5 times in case the Update API is busy; Waiting 10 seconds between each attempt.
+	for attempts < 5 {
```

Comment:
It looks like attempts is never incremented.

```suggestion
	for ; attempts < 5; attempts++ {
```